webservice/models/webservice_backend.py

```diff
@@ -71,8 +82,40 @@ def _valid_field_parameter(self, field, name):
         extra_params = ("auth_type",)
         return name in extra_params or super()._valid_field_parameter(field, name)
 
-    def call(self, method, *args, **kwargs):
-        return getattr(self._get_adapter(), method)(*args, **kwargs)
+    @contextmanager
+    def _consumer_record_env(self, record, new_cursor=True):
+        if new_cursor:
+            with api.Environment.manage():
+                with registry(self.env.cr.dbname).cursor() as new_cr:
+                    new_env = api.Environment(new_cr, self.env.uid, self.env.context)
+                    # in case of error the main transaction will be rollback
+                    # in any case we want to save the response payload for
+                    # later analysis. Some webservice gives informations regarding
+                    # the cause of failures
+                    yield record.with_env(new_env)
+        else:
+            yield record
+
+    def call(self, method, *args, consumer_record=None, new_cursor=True, **kwargs):
+        content = False
+        status_code = False
+        try:
+            content = getattr(self._get_adapter(), method)(*args, **kwargs)
+            status_code = 200
+        except HTTPError as request_error:
+            content = request_error.response.content
+            status_code = request_error.response.status_code
+            raise request_error from request_error
+        finally:
+            if self.save_response and consumer_record:
+                with self._consumer_record_env(
```

Comment:
IMO the new_cursor flag shouldn't be passed to `call`, it should be _determined_ by call.
If you catch the exception you can set the flag to true and pass it to the ctx manager.
Also, when you get an HTTPError you raise immediately instead of storing that response too. 
This would solve the problem no?

Reply:
if I create a new transaction only in case of failure that will definitively works... expect if code is raising later in the stack the response won't be saved, this is kind of okay in my use case.

Saving response only as ir.attachement would solve the problem too but without back ref on the consumer records it will be hard to reuse the response later so I'll go for the first suggestion, creating new transaction only in case of failure !

Reply:
I see your point but probably this way is more consistent?
As an alternative we could use a savepoint later...